python/cudf/cudf/tests/dataset_generator.py

```diff
@@ -0,0 +1,210 @@
+            safe=False,
+        )
+
+    else:
+        # Generate data for current column
+        return pa.array(
+            column_params.generator,
+            mask=np.random.choice(
+                [True, False],
+                size=num_rows,
+                p=[
+                    column_params.null_frequency,
+                    1 - column_params.null_frequency,
+                ],
+            )
+            if column_params.null_frequency > 0.0
+            else None,
+            size=num_rows,
+            safe=False,
+        )
+
+
+def generate(
+    path,
+    parameters,
+    format={"name": "parquet", "row_group_size": 64},
+    use_threads=True,
+):
+    """
+    Generate dataset using given parameters and write to given format
+
+    Parameters
+    ----------
+    path : str or file-like object
+        Path to write to
+    parameters : Parameters
+        Parameters specifying how to randomly generate data
+    format : Dict
+        Format to write
+    """
+
+    # Initialize seeds
+    if parameters.seed is not None:
+        np.random.seed(parameters.seed)
+    column_seeds = np.arange(len(parameters.column_parameters))
+    np.random.shuffle(column_seeds)
+
+    # For each column, use a generic Mimesis producer to create an Iterable
+    # for generating data
+    for i, column_params in enumerate(parameters.column_parameters):
+        column_params.generator = column_params.generator(
+            Generic("en", seed=column_seeds[i])
+        )
+
+    # Get schema for each column
+    schema = pa.schema(
+        [
+            pa.field(
+                name=str(i),
+                type=pa.from_numpy_dtype(
+                    type(next(iter(column_params.generator)))
+                ),
+                nullable=column_params.null_frequency > 0,
+            )
+            for i, column_params in enumerate(parameters.column_parameters)
+        ]
+    )
+
+    # Initialize column data and which columns should be sorted
+    column_data = [None] * len(parameters.column_parameters)
+    columns_to_sort = [
+        str(i)
+        for i, column_params in enumerate(parameters.column_parameters)
+        if column_params.is_sorted
+    ]
+
+    # Generate data
+    if not use_threads:
+        for i, column_params in enumerate(parameters.column_parameters):
+            column_data[i] = _generate_column(
+                column_params, parameters.num_rows
+            )
+    else:
+        pool = Pool(pa.cpu_count())
+        column_data = pool.starmap(
+            _generate_column,
+            [
+                (column_params, parameters.num_rows)
+                for i, column_params in enumerate(parameters.column_parameters)
+            ],
+        )
+        pool.close()
+
+    # Convert to Pandas DataFrame and sort columns appropriately
+    tbl = pa.Table.from_arrays(column_data, schema=schema,)
+    if columns_to_sort:
+        tbl = tbl.to_pandas()
+        tbl = tbl.sort_values(columns_to_sort)
```

Comment:
We'll lose all of our nulls in integer columns here which we don't want to do.

Idea: keep the pyarrow table around, create a Panda dataframe from the pyarrow table, set the columns you want to sort by as the index, run an argsort against the index to get a gathermap, use the gathermap to gather from the PyArrow table:
```
tbl = pa.Table.from_arrays(column_data, schema=schema,)
if columns_to_sort:
    pdf = tbl.to_pandas()
    pdf = pdf.set_index(columns_to_sort)
    gather_map = pdf.index.argsort()
    tbl = tbl.take(gather_map)
```

Reply:
> We'll lose all of our nulls in integer columns here which we don't want to do.

I'm not able to reproduce this. The following script should be generating integers with a null mask applied. And this does sorting so the `columns_to_sort` path should be taken.
```python
dg.generate(
    fname,
    dg.Parameters(
        num_rows=2048,
        column_parameters=[
            dg.ColumnParameters(
                cardinality=40,
                null_frequency=0.05,
                generator=lambda g: [g.address.city() for _ in range(40)],
                is_sorted=False,
            ),
            dg.ColumnParameters(
                40,
                0.2,
                lambda g: [g.person.age() for _ in range(40)],
                True,
            ),
        ],
        seed=rdg_seed,
    ),
    format={"name": "parquet", "row_group_size": 64},
)
```
And I'm getting the output sorted with what I think are nulls.
```python
In [6]: pd.read_parquet("simple.parquet")
Out[6]: 
                   0     1
1290     Walla Walla  18.0
885      Walla Walla  18.0
52       Walla Walla  18.0
1845     Walla Walla  18.0
446      Walla Walla  18.0
...              ...   ...
2014         Antioch   NaN
2015            None   NaN
2021            None   NaN
2036            None   NaN
2043  Tarpon Springs   NaN

[2048 rows x 2 columns]
```
If the underlying issue is that we can't depend on Pandas for writing data correctly, I rewrote the code to depend on PyArrow-to-Pandas and Pandas-to-PyArrow conversion (overhead of sorting is 3s instead of 2.5s for 2 million rows of low-entropy data with above config) working correctly with the resulting PyArrow table being written to Parquet. This looks like the following.
```python
    tbl = pa.Table.from_arrays(column_data, schema=schema,)
    if columns_to_sort:
        tbl = tbl.to_pandas()
        tbl = tbl.sort_values(columns_to_sort)
        tbl = pa.Table.from_pandas(tbl)
```

I added a TODO to use your solution once #5835 is resolved (I think @trxcllnt was working on upgrading now that arrow-cpp 1.0.1 is out [1]) and we can use `pyarrow.compute.take`.

[1] https://github.com/trxcllnt/cudf/tree/fea/arrow-1.0.0

Reply:
You're only getting nulls in the String dtype columns as Pandas uses Python objects under the covers for those anyway. For integers with nulls, calling `to_pandas()` is going to cast them to a float dtype column with `NaN` values instead which is what I was trying to avoid.

From my testing locally, it seems like PyArrow will automatically replace the `NaN` values with `null` values, but you'll still end up going from `int64` --> `float64` for example, but we could address that with some casting on the PyArrow side?

Reply:
OK, I understand now. `to_pandas` does have an `integer_object_nulls` parameters but when I set it to true, I'm still getting the same result above - `float64` mixed with `NaN`.

Reply:
I'm saying that we don't have to wait, where we can do something like:
```
    tbl = pa.Table.from_arrays(column_data, schema=schema,)
    if columns_to_sort:
        tbl = tbl.to_pandas().sort_values(columns_to_sort)
        tbl = pa.Table.from_pandas(tbl).cast(schema)
```

The last line will cast the `intXX` --> `floatXX` converted columns back to `intXX` dtype as it will use the original schema that we used to build the initial PyArrow Table.

Reply:
I see, I didn't actually check to see PyArrow lets you do that. Thanks. The following change fixes the issue - nullable data sorted by Pandas gets cast back into its original data type and I confirmed that by reading back in the generated Parquet with PyArrow.
```python
tbl = pa.Table.from_pandas(tbl, schema)
```

Reply:
@kkraus14 I believe this is resolved now with https://github.com/rapidsai/cudf/pull/5843/commits/5ccb78e69077508045c38f8ed0d92ed6cf0f10ed?